File: .github/workflows/validate_workflows.py
#!/usr/bin/env python3
"""Validate archon workflow YAML files."""
import sys, yaml
from pathlib import Path

WORKFLOWS_DIR = Path(__file__).parents[2] / '.archon' / 'workflows'
REQUIRED_TOP_FIELDS = {'name', 'description'}
VALID_NODE_TYPES = {'command', 'prompt', 'bash', 'loop'}
VALID_COMMANDS = {
    'analyze-dependencies', 'auto-fix', 'auto-update', 'build',
    'bump-version', 'check-breaking-changes', 'check-coverage',
    'check-env-file', 'check-gate', 'check-outdated', 'check-prerequisites',
    'code-scan', 'commit-push', 'cost-report', 'create-issue',
    'create-release', 'create-tag', 'dependency-audit', 'deploy',
    'detect-language', 'docs-update', 'generate-changelog', 'generate-env',
    'generate-report', 'git-leaks', 'install-deps', 'knip-check',
    'list-releases', 'memory-recall', 'memory-remember', 'notify-complete',
    'opendigger-analyze', 'performance-profile', 'production-build',
    'review-full', 'rollback', 'run-suite', 'sec-bug-hunt',
    'start-dev-server', 'test-coverage', 'test-unit', 'update-deps', 'verify',
}

def validate_node_chain(nodes):
    node_ids = {n['id'] for n in nodes}
    deps_from = set()
    deps_to = set()
    for n in nodes:
        for d in n.get('depends_on', []):
            if d not in node_ids:
                print(f"  [FAIL] Node '{n['id']}' depends on unknown node '{d}'")
                return False
            deps_from.add(n['id'])
            deps_to.add(d)
    entries = node_ids - deps_from
    terminals = node_ids - deps_to
    if not entries:
        print(f"  [FAIL] No entry node found (circular dependency?)")
        return False
    if not terminals:
        print(f"  [FAIL] No terminal node found")
        return False
    return True

def validate_workflow(filepath):
    name = filepath.name
    print(f"Validating {name}...")
    with open(filepath, 'r', encoding='utf-8') as f:
        try:
            data = yaml.safe_load(f)
        except yaml.YAMLError as e:
            print(f"  [FAIL] YAML syntax error: {e}")
            return False
    if not isinstance(data, dict):
        print(f"  [FAIL] Not a mapping")
        return False
    for field in REQUIRED_TOP_FIELDS:
        if field not in data:
            print(f"  [FAIL] Missing required field: {field}")
            return False
    if 'nodes' not in data:
        print(f"  [FAIL] Missing 'nodes' section")
        return False
    node_ids = set()
    for node in data['nodes']:
        nid = node.get('id')
        if not nid:
            print(f"  [FAIL] Node missing 'id'")
            return False
        if nid in node_ids:
            print(f"  [FAIL] Duplicate node ID: {nid}")
            return False
        node_ids.add(nid)
        ntype = node.get('type', 'command')
        if ntype not in VALID_NODE_TYPES:
            print(f"  [FAIL] Node '{nid}' has invalid type '{ntype}'")
            return False
        if 'description' not in node:
            print(f"  [WARN]  Node '{nid}' missing description")
        cmd = node.get('command')
        if cmd and cmd not in VALID_COMMANDS:
            print(f"  [WARN]  Node '{nid}' uses unregistered command '{cmd}'")
    if not validate_node_chain(data['nodes']):
        return False
    print(f"  [OK] {len(data['nodes'])} nodes, valid")
    return True

def main():
    yamls = sorted(WORKFLOWS_DIR.glob('*.yaml'))
    if not yamls:
        print("No YAML files found!")
        sys.exit(1)
    ok = 0
    for y in yamls:
        if validate_workflow(y):
            ok += 1
    print(f"\nResults: {ok}/{len(yamls)} passed")
    sys.exit(0 if ok == len(yamls) else 1)

if __name__ == '__main__':
    main()

File: .github/workflows/validate_workflows.js
#!/usr/bin/env node
/**
 * archon Workflow Validator (Node.js version)
 * Validates .archon/workflows/*.yaml structure.
 */

const fs = require('fs');
const path = require('path');

// Actual node types used in workflows
const VALID_NODE_KEYS = new Set(['command', 'prompt']);
const REQUIRED_FIELDS = ['id', 'description'];

function validateWorkflow(filepath) {
  console.log(`Validating ${path.basename(filepath)}...`);

  try {
    const content = fs.readFileSync(filepath, 'utf8');
    const lines = content.split('\n');

    const nodeIds = new Set();
    const dependencyMap = new Map(); // nodeId -> [depIds]
    let hasErrors = false;
    let currentNode = null;
    let inPrompt = false;

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      const trimmed = line.trim();
      if (!trimmed) continue;

      const indent = line.length - line.trimStart().length;

      // Detect node start: "- id: xxx" at indent 2
      if (indent === 2 && /^- id:/.test(trimmed)) {
        // Validate previous node
        if (currentNode && currentNode.id) {
          const missing = REQUIRED_FIELDS.filter(f => !currentNode[f]);
          if (missing.length) {
            console.log(`  ❌ Node '${currentNode.id}' missing: ${missing.join(', ')}`);
            hasErrors = true;
          }
          if (currentNode.id) nodeIds.add(currentNode.id);
        }

        currentNode = {
          id: trimmed.replace(/^- id:\s*/, '').replace(/^["']|["']$/g, '')
        };
        dependencyMap.set(currentNode.id, []);
        inPrompt = false;
      }
      else if (currentNode && indent === 4 && !inPrompt) {
        // Parse node fields (indent 4)
        if (trimmed.startsWith('description:')) {
          currentNode.description = trimmed.replace(/^description:\s*/, '');
        } else if (trimmed.startsWith('command:')) {
          currentNode.command = trimmed.replace(/^command:\s*/, '');
          currentNode.type = 'command';
        } else if (trimmed.startsWith('prompt:')) {
          currentNode.type = 'prompt';
          inPrompt = true; // subsequent indented lines are prompt body
        } else if (trimmed.startsWith('depends_on:')) {
          const depStr = trimmed.replace(/^depends_on:\s*/, '');
          const deps = depStr.replace(/^\[|\]$/g, '').split(',').map(d => d.trim().replace(/^["']|["']$/g, '')).filter(Boolean);
          dependencyMap.set(currentNode.id, deps);
        }
      }
      // Lines indented > 4 inside a prompt node are prompt body, skip
    }

    // Validate last node
    if (currentNode && currentNode.id) {
      const missing = REQUIRED_FIELDS.filter(f => !currentNode[f]);
      if (missing.length) {
        console.log(`  ❌ Node '${currentNode.id}' missing: ${missing.join(', ')}`);
        hasErrors = true;
      }
      if (currentNode.id) nodeIds.add(currentNode.id);
    }

    if (nodeIds.size === 0) {
      console.log('  ❌ No valid nodes found');
      return false;
    }

    // Check depends_on references
    for (const [nodeId, deps] of dependencyMap) {
      for (const dep of deps) {
        if (dep && !nodeIds.has(dep)) {
          console.log(`  ❌ Broken dependency: '${nodeId}' depends on non-existent '${dep}'`);
          hasErrors = true;
        }
      }
    }

    // Check duplicate IDs
    const allIdLines = lines
      .map((l, i) => ({ line: l.trim(), idx: i + 1 }))
      .filter(l => /^- id:/.test(l.line));
    const seen = new Map();
    for (const { line, idx } of allIdLines) {
      const id = line.replace(/^- id:\s*/, '');
      if (seen.has(id)) {
        console.log(`  ❌ Duplicate node ID '${id}' at line ${idx} (first at line ${seen.get(id)})`);
        hasErrors = true;
      } else {
        seen.set(id, idx);
      }
    }

    if (!hasErrors) {
      console.log(`  ✅ ${path.basename(filepath)} is valid (${nodeIds.size} nodes)`);
      return true;
    }
    return false;
  } catch (err) {
    console.log(`  ❌ Error: ${err.message}`);
    return false;
  }
}

function main() {
  const workflowsDir = path.join(process.cwd(), '.archon', 'workflows');

  if (!fs.existsSync(workflowsDir)) {
    console.error(`❌ Directory not found: ${workflowsDir}`);
    process.exit(1);
  }

  const files = fs.readdirSync(workflowsDir).filter(f => f.endsWith('.yaml'));
  console.log(`Found ${files.length} workflow files\n`);

  let allValid = true;
  for (const file of files) {
    if (!validateWorkflow(path.join(workflowsDir, file))) {
      allValid = false;
    }
  }

  console.log();
  if (allValid) {
    console.log('✅ All workflows are valid');
    process.exit(0);
  } else {
    console.log('❌ Some workflows have errors');
    process.exit(1);
  }
}

main();
